Utilities/SharedMutex.h

```diff
@@ -1,46 +1,136 @@
 #pragma once
 
-#include <shared_mutex>
-
-// An attempt to create lock-free (in optimistic case) implementation similar to std::shared_mutex;
-// MSVC implementation of std::shared_timed_mutex is not lock-free and thus may be slow, and std::shared_mutex is not available.
-class shared_mutex_t
+//! An attempt to create effective implementation of "shared mutex", lock-free in optimistic case.
+//! All locking and unlocking may be done by single LOCK XADD or LOCK CMPXCHG instructions.
+//! MSVC implementation of std::shared_timed_mutex seems suboptimal.
+//! std::shared_mutex is not available until C++17.
```

Comment:
Is this shared_mutex API compatible with the std one ? ie Does std::unique_lock/ std::lock_guard compatible with it ?
I think it should be precised in this comment.